airflow/models/dag.py

```diff
@@ -2867,8 +2877,45 @@ def bulk_write_to_db(
             task_refs_stored = existing_task_outlet_refs_dict[(dag_id, task_id)]
             task_refs_to_add = {x for x in task_refs_needed if x not in task_refs_stored}
             session.bulk_save_objects(task_refs_to_add)
-            for obj in task_refs_stored - task_refs_needed:
-                session.delete(obj)
+
+            task_refs_to_remove = {x for x in task_refs_stored if x not in task_refs_needed}
+            for todr in task_refs_to_remove:
+                log.debug(
+                    "Removing TaskOutletDatasetReference (dataset_id=%s, dag_id=%s, task_id=%s)",
+                    todr.dataset_id,
+                    todr.dag_id,
+                    todr.task_id,
+                )
+                session.query(TaskOutletDatasetReference).filter(
+                    TaskOutletDatasetReference.dataset_id == todr.dataset_id,
+                    TaskOutletDatasetReference.dag_id == todr.dag_id,
+                    TaskOutletDatasetReference.task_id == todr.task_id,
+                ).delete()
+
+        # Remove any orphaned datasets
+        orphaned_dataset_query = (
```

Comment:
This looks pretty good, but we might want it do it like we do for stale DAGs: https://github.com/apache/airflow/blob/093345cb61cea623bffe6af5a41d6ba417dfcf6f/airflow/dag_processing/manager.py#L492